File: library-management/csharp/src/LibraryManagement/Loan.cs
namespace LibraryManagement;

public class Loan
{
    public const int LoanPeriodDays = 14;
    public static readonly Money FinePerDay = new(0.10m);

    public Loan(Member member, Copy copy, DateOnly borrowedOn)
    {
        Member = member;
        Copy = copy;
        BorrowedOn = borrowedOn;
        DueOn = borrowedOn.AddDays(LoanPeriodDays);
    }

    public Member Member { get; }
    public Copy Copy { get; }
    public DateOnly BorrowedOn { get; }
    public DateOnly DueOn { get; }
    public DateOnly? ReturnedOn { get; private set; }

    public bool IsClosed => ReturnedOn is not null;

    public Money FineFor(DateOnly returnDate)
    {
        if (returnDate <= DueOn) return Money.Zero;
        var daysLate = returnDate.DayNumber - DueOn.DayNumber;
        return FinePerDay * daysLate;
    }

    internal void Close(DateOnly returnedOn) => ReturnedOn = returnedOn;
}

File: library-management/typescript/src/Loan.ts
import { Copy } from './Copy.js';
import { Member } from './Member.js';
import { Money } from './Money.js';

export const LOAN_PERIOD_DAYS = 14;
export const FINE_PER_DAY = new Money(0.10);

const MS_PER_DAY = 86_400_000;

export class Loan {
  public readonly dueOn: Date;
  private _returnedOn: Date | null = null;

  constructor(
    public readonly member: Member,
    public readonly copy: Copy,
    public readonly borrowedOn: Date,
  ) {
    this.dueOn = new Date(borrowedOn.getTime() + LOAN_PERIOD_DAYS * MS_PER_DAY);
  }

  get returnedOn(): Date | null { return this._returnedOn; }
  get isClosed(): boolean { return this._returnedOn !== null; }

  fineFor(returnDate: Date): Money {
    if (returnDate.getTime() <= this.dueOn.getTime()) return Money.zero;
    const daysLate = Math.round((returnDate.getTime() - this.dueOn.getTime()) / MS_PER_DAY);
    return FINE_PER_DAY.times(daysLate);
  }

  close(returnedOn: Date): void { this._returnedOn = returnedOn; }
}
